packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -595,9 +596,15 @@ function updateDynamicChildren(parentElm: Node, oldCh: VNodes, newCh: VNodes) {
                         // Delete the old child, but copy the array since it is read-only.
                         // The `oldCh` will be GC'ed after `updateDynamicChildren` is complete,
                         // so we only care about the `oldCh` object inside this function.
-                        const oldChClone = [...oldCh];
-                        oldChClone[idxInOld] = undefined as any;
-                        oldCh = oldChClone;
+                        // To avoid cloning over and over again, we check `clonedOldCh`
+                        // and only clone once.
+                        if (!clonedOldCh) {
+                            clonedOldCh = true;
+                            oldCh = [...oldCh];
+                        }
```

Comment:
I considered doing an object equality test here (e.g. create a local variable assigned to `oldCh`, then later check if `oldCh === clonedOldCh`), but I couldn't think of good names for the variables, so I figured this approach was better.